### PR TITLE
Performance improvements for the v2 analyzer driver

### PR DESCRIPTION
**Pull request by mavasani**
Fix performance regression in the IDE diagnostic analyzer service. 

1) Reduce number allocations in the analyzer driver that show up perf traces for Picasso.
2) Currently we maintain a cache of CompilationWithAnalyzers for projects being actively analyzed. This is causing the VM to rise and also forcing GC to kick in at different phases causing regressions at various scenarios. The fix is to just create a new CompilationCompilationWithAnalyzers instance for each document/project analysis request.

### CODE REVIEW

**Comment by mavasani** [comment]
@dotnet-bot retest this please.

**Comment by mavasani** [comment]
@heejaechang @JohnHamby @pharring @KevinH-MS can you please take a look? This should bring the Open file/typing perf close to back where it was before the v2 driver change (see private RPS run on 23302.00)

**Comment by mavasani** [comment]
@dotnet-bot retest this please

**Comment by mavasani** [comment]
FYI: I am going to merge this in today to get new performance numbers over the weekend, and if required, perform any additional investigations. I'll address PR feedback next week.

**Comment by heejaechang** [comment]
:+1: it is sad that we have to do this for allocations... which I have done many times :)

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.CompilationData.cs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal abstract partial class AnalyzerDriver : IDisposable
+    {
+        protected static readonly ConditionalWeakTable<Compilation, CompilationData> s_compilationDataCache = new ConditionalWeakTable<Compilation, CompilationData>();
```
Comment (pharring) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.CompilationData.cs]: Now would be a good time to add a general-purpose property bag to Compilation and move all the ```ConditionalWeakTable<Compliation, ...>``` instances we have into that.

Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.CompilationData.cs]: +1


-          Heejae

**Pull request merged**